### PR TITLE
feat(client): add `transactionOptions` to `PrismaClient`

### PR DESCRIPTION
**Pull request by tockn**
Thank you for the greate project!
I have implemented a feature that allows setting default values for `Transaction` options, such as `maxWait` , `timeout` , and `isolationLevel` , during the initialization of `PrismaClient` . Previously, it was difficult to unify the default values for `Transaction` created from a specific `PrismaClient` because these options could only be passed during the instantiation of a `Transaction` instance.

With this enhancement, you can now configure the default values for `Transaction` options when creating a `PrismaClient` . I believe this feature will be beneficial, and I would greatly appreciate your review if you find it valuable.

### CODE REVIEW

**Comment by tockn** [comment]
@SevInf @aqrln Hi! Thank you for maintaining it!
Is there any updates?

**Review by aqrln — COMMENTED** [review]
The changes look good, but the new API needs to be tested. Would you be able to add a test for this, or do you need any assistance with it?

**Comment by Jolg42** [comment]
@tockn did you see this message maybe?

**Comment by michaelhays** [comment]
This would be a very welcome change -- we use a 20 second timeout for all of our transactions, and would love to be able to set that by default.

**Comment by vpishuk** [comment]
Any chance this can be merged soon?

**Comment by tockn** [comment]
Sorry for the delay in responding! I will respond shortly!

**Comment by tockn** [comment]
@aqrln 
I have implemented the TSClient and added tests. Could you review it again?

**Comment by tockn** [comment]
@SevInf @aqrln Is there any updates or anything else I should do?

**Comment by tockn** [comment]
@aqrln @SevInf Hi! I want to know the situation. Is there anything I should do?

**Comment by shrujalshah28** [comment]
@Jolg42 Any update on this? Anything I can help.

**Comment by Jolg42** [comment]
@shrujalshah28 I'm not sure what's the current state at the moment, I remember that his was quite a nice contribution.

At the moment, to move this forward, we would to fix the git conflicts and review it.

**Comment by shrujalshah28** [comment]
@Jolg42 I have opened MR for this - https://github.com/prisma/prisma/pull/22583

**Comment by tockn** [comment]
Thank you. I will fix the conflicts!

**Comment by tockn** [comment]
@Jolg42 Thank you for your response! I have resolved the conflicts. Please review it! (Some test cases have failed, but they seem like flaky tests?)

**Comment by tockn** [comment]
@Jolg42 Would it be possible for you to review this for me?

**Comment by Jolg42** [comment]
@tockn Sorry for the delay, I didn't forget you but didn't have the time as I had more urgent things to do. I'll try to take a moment this week and report back.

File: packages/client/src/generation/TSClient/PrismaClient.ts [hunk 1]
```diff
@@ -544,6 +544,21 @@ export type TransactionClient = Omit<Prisma.DefaultPrismaClient, runtime.ITXClie
...
           `),
       )
+      .add(
+        ts
+          .property(
+            'transactionOptions',
+            ts
+              .objectType()
+              .add(ts.property('maxWait', ts.numberType).optional())
+              .add(ts.property('timeout', ts.numberType).optional()),
```
Comment (SevInf) [inline packages/client/src/generation/TSClient/PrismaClient.ts:554]: This type does not include isolation level, where it runtime, the property exists and affects the execution. You can see how to do add isolation level property [here](https://github.com/prisma/prisma/blob/d212cd87d0e3586fde996f2c479226a3d53b22e1/packages/client/src/generation/TSClient/PrismaClient.ts#L174)

File: packages/client/src/runtime/core/engines/binary/BinaryEngine.ts [hunk 2]
```diff
@@ -812,9 +812,9 @@ You very likely have the wrong "binaryTarget" defined in the schema.prisma file.
 
     if (action === 'start') {
       const jsonOptions = JSON.stringify({
-        max_wait: arg?.maxWait ?? 2000, // default
-        timeout: arg?.timeout ?? 5000, // default
-        isolation_level: arg?.isolationLevel,
+        max_wait: arg?.maxWait ?? this.config.transactionOptions.maxWait,
```
Comment (SevInf) [inline packages/client/src/runtime/core/engines/binary/BinaryEngine.ts:815]: Rather then copying this code into 3 engine classes, I'd suggest to do it in a single place in [`getPrismaClient`](https://github.com/prisma/prisma/blob/d212cd87d0e3586fde996f2c479226a3d53b22e1/packages/client/src/runtime/getPrismaClient.ts#L710).

Additionally, default isolation level should probably also apply to batch transactions (the ones that take in array), where this path handles only iTx (the ones that take callback).

The place to apply default to the batch transactions is [a couple of lines above](https://github.com/prisma/prisma/blob/d212cd87d0e3586fde996f2c479226a3d53b22e1/packages/client/src/runtime/getPrismaClient.ts#L689)

Comment (tockn) [reply packages/client/src/runtime/core/engines/binary/BinaryEngine.ts]: maybe fixed by those commits.
https://github.com/prisma/prisma/pull/19592/commits/be405daa9315b71a51c55f987e85899176eef0fe
https://github.com/prisma/prisma/pull/19592/commits/e855ca6c05581e4e2069aba94fc9820105ce5a8d

**Comment by tockn** [comment]
@SevInf thank you for reviewing! i've fixed those points!

**Review by Jolg42 — APPROVED** [review]
Nice work!

**Review by SevInf — APPROVED** [review]

**Comment by Jolg42** [comment]
@tockn (and others) you can (finally!) try this out in the following dev version `5.10.0-dev.71`.

Thanks again for the good work 💚 

Let us know if you have any feedback 🙌🏼 

I recommend waiting for the official `5.10.0` release scheduled, next week, on Tuesday, 20th February to use it in production.

**Pull request merged**